config/confighttp/confighttp.go

```diff
@@ -242,12 +245,31 @@ func (hcs *ClientConfig) ToClient(ctx context.Context, host component.Host, sett
 	}
 
 	return &http.Client{
-		Transport: clientTransport,
-		Timeout:   hcs.Timeout,
-		Jar:       jar,
+		CheckRedirect: makeCheckRedirect(hcs.MaxRedirects),
+		Transport:     clientTransport,
+		Timeout:       hcs.Timeout,
+		Jar:           jar,
 	}, nil
 }
 
+// makeCheckRedirect checks if max redirects are exceeded
+func makeCheckRedirect(max *int) func(*http.Request, []*http.Request) error {
+	if max == nil {
+		return nil
+	} else if *max == 0 {
```

Comment:
Is this case needed here? if max is set to 0, will the check at line 266 return on the first call the same as is this block here?

Reply:
Good point, the test case is needed because the length of the `via []*http.Request` variable on line 265 will be at least 1. This is because the `CheckRedirect` function of the standard htp client is always called after getting a redirect http response, so at least one request has been made:

```
	// If CheckRedirect is not nil, the client calls it before
	// following an HTTP redirect. The arguments req and via are
	// the upcoming request and the requests made already, oldest
	// first.
``` 

If you remove this conditional branch, the "Zero redirects" test case will fail (never end) as the length check will never match (number of already done request will be always > 0).

Reply:
Great catch! Actually, this branch is not needed and resulted in the same output when `max_redirects` was set to either 0 or 1. The max redirects function is always called when a redirect response is returned, meaning that at least one request has already been made. The variable `via []*http.Request` on line 265 will always be > 0.

If `max_redirects` is set to 0, the http client should not follow any redirect response (one request). If `max_redirects` is 1, only one redirect response should be followed (2 total requests). `max_redirects = total_requests - 1`

Fixed in https://github.com/open-telemetry/opentelemetry-collector/pull/10877/commits/c98fe0a5fcffcc1994e58db23c5077e02cb2997c

Thanks!